docs/design/factories.md

```diff
@@ -0,0 +1,234 @@
+which lets the framework call the function to instantiate a new instance of the
+type. This works whether the type is defined by the provider (the function
+would return an interface) or by the framework.
+
+### Named functions
+
+Similarly to anonymous functions, we can ask for named functions:
+
+```go
+type ResourceFactory func() framework.Resource
+
+type Provider struct {
+  Resources map[string]ResourceFactory
+}
+```
+
+This also asks for a function that returns a `framework.Resource`, but gives
+the function signature a name. This works whether the type is defined by the
+provider (the function would return an interface) or by the framework.
+
+### Reflection
+
+In theory, given a single instance of the type, we can use the `reflect`
+package to create new instances of the type:
+
+
+```go
+resource := computeInstanceResource{}
+
+typ := reflect.TypeOf(resource)
+newResource := reflect.New(typ)
+
+// newResource is now a newly instantiated variable of the same type as
+// resource
+```
+
+We can use this in a helper, to do something like:
+
+```go
+type Provider struct {
+  Resources map[string]Resource
+}
+```
+
+and do the instantiation ourselves behind the scenes.
+
+This works if the type is defined by the provider, but there's not really any
+point to it if the type is defined by the framework, as the framework can just
+use manual copying to get the same result.
+
+### Manual copying
+
+If the Resource type is owned by the framework, it will be able to instantiate
+its own instances of that type, and would be able to copy over the data used to
+populate them:
+
+```go
+resource := framework.Resource{
+	Schema: map[string]*Schema{},
+	Create: createFunc,
+	Read: readFunc,
+	Update: updateFunc,
+	Delete: deleteFunc,
+}
+
+newResource := framework.Resource{}
+newResource.Schema = resource.Schema
+newResource.Create = resource.Create
+newResource.Read = resource.Read
+newResource.Update = resource.Update
+newResource.Delete = resource.Delete
+```
+
+we can use this in a helper, to do something like:
+
+```go
+type Provider struct {
+	Resources map[string]Resource
+}
+```
+
+and do the instantiation ourselves behind the scenes.
+
+This works only if the type is defined by the framework. Types defined by the
+provider would need to use reflection to achieve this outcome.
+
+### Factory types
+
+Instead of using functions, we can have an interface for a factory type that
+providers define:
+
+```go
+type ResourceFactory interface {
+  NewResource() Resource
+}
+```
+
+Which can then be used instead of the type the factory will return when asking
+users for resources, data sources, and providers:
+
+```go
+type Provider struct {
+  Resources map[string]ResourceFactory
+}
+```
+
+This works whether the type is defined by the provider (the `NewResource()`
+method would return an interface) or by the framework.
+
+### Separating types and values
+
+The SDK currently conflates two separate ideas: the type of a resource--i.e.,
```

Comment:
I'm not sure I understand the conflation in question here. The SDK currently defines a `schema.Resource` type, which is a struct:
```go
type Resource struct {
  Schema map[string]*Schema
  // all our familiar resource fields
}
```

This is a type. 

In provider code, instances of this type are created:
```go
myResource := &schema.Resource{
  Schema : ...
  // etc.
}
```

The SDK has no trouble distinguishing these two things, because one is a type and one is a value, just as the `http.Client` type is different from `&http.Client{}`.

When the SDK needs to talk about _a single resource definition in state_, it has its own types for this: `terraform.ResourceState` and `terraform.InstanceState`. As far as I can tell, provider code is not concerned with these, because the job of the provider server is to:
1. Receive an RPC specifying a certain `type_name` of resource (e.g. `compute_instance`) and the config and state for that resource instance;
2. Perform API requests as necessary to make sure the state of the world matches the Terraform state;
3. Respond to the RPC with the updated state values.

I haven't found any examples of idiomatic Go that look like the `ResourceType` and `ResourceValue` constructs below, and it's unclear to me how they map to the problem of separating _the general idea of a resource_ from _a single resource definition in state_. A resource's `Create` method is always the same regardless of which instance of the resource we're talking about - that information comes in via the _parameters_ of the `Create` function.

Reply:
I think I may have accidentally confused the matter by (:sadtrombone:) using the words "type" and "value", so I'm gonna take a stab at clarifying that.

Essentially, I think resources are used in two different ways, and the best way to talk about them is from the perspective of the framework.

The first way the framework sees resources is as a source of schemas. Technically, this doesn't even _need_ to be part of the Go type implementing the resource (it could just be all defined inline in the provider GetSchema method), but I think conceptually it is part of a resource. This can be not-accurately-but-not-entirely-inaccurately as a call that will never happen in parallel to other GetSchema calls. I think, therefore, it makes sense to think of whatever is implementing the GetSchema functionality can be thought of as a singleton-per-resource-per-provider-instance. You're getting the schema of what I'm calling the "resource type" (not to be confused with the framework.Resource type), and there's no real reason to isolate these.

The second way the protocol sees resources is as a group of methods that will be called at runtime. These _will_ be run in parallel, likely, and are concerned with a single _instance_ of the resource. Yes, the methods will stay the same, but the context they're called in will be specific to the instance of a resource, where as the context of GetSchema is called in the context of all instances of the resource. I think, therefore, it makes sense to think about implementing these as "values" or "instances" of that resource "type" and providing isolation between them.

I think, in code, what I'm talking about could be described like this:

```go
// inside the framework module
type ResourceType struct {
  Schema map[string]Schema
  NewResource func() Resource
}

type Resource interface {
  Create(ctx context.Context) // etc
}
```

```go
// inside the provider

// the resource "type" is just an instance of a framework.ResourceType type. All resource types share the same
// implementation/functionality, the only data that changes is the schema, the shape of the resource.
var computeInstanceType = framework.Resource{
  Schema: map[string]framework.Schema{
    "name": framework.Schema{},
  },
  NewResource: func() framework.Resource{
    return computeInstanceResource{}
  }
}

// the CRUD methods are defined on a separate type, because they aren't shared among all resource types, they're shared
// among all instances of a resource type.
type computeInstanceResource struct{}

func (c computeInstanceResource) Create(ctx context.Context) {}
```

Basically, I think if we're thinking of Go types as a collection of common data structures and behaviors, "things that return schemas" have more in common with each other than they do with "things that operate on data described by their schema", given how they're used.

This is related to the factory problem because I discovered this while trying out things like this:

```go
type Resource interface{
  GetSchema()
  NewResource() Resource
  Create()
  Read()
  Update()
  Delete()
}
```

Basically, combining what I'm calling "ResourceType" and "Resource"/"ResourceValue" into a single type. In that situation, we're stuck in a loop--we need an instance of `Resource` to get an instance of `Resource`. This led to factories:

```go
type ResourceFactory interface {
  NewResource() Resource
}

type Resource interface{
  GetSchema()
  Create()
  Read()
  Update()
  Delete()
}
```

But having a separate type that doesn't hold any data doesn't make sense. And then I realised there was no reason it _couldn't_ hold the schema, and that draws a nice separation between the Schema as something Terraform retrieves at the beginning of the run and the CRUD calls as something that happens on instances of that Schema.

I'm really struggling to articulate this, sorry :(

Reply:
Yes, let's at least rephrase this so we're not using the overloaded terms "type" and "value" - I still think the distinction between `ResourceType` and `Resource`/`ResourceValue` is confusing and doesn't do much to help provider developers.


> Yes, the methods will stay the same, but the context they're called in will be specific to the instance of a resource, where as the context of GetSchema is called in the context of all instances of the resource. 

This is what I meant by "A resource's Create method is always the same regardless of which instance of the resource we're talking about - that information comes in via the parameters of the Create function." - this distinction is already incorporated in the parameters of `GetSchema` vs those of `Create`. Providers do not need to have the concept of resource instance exposed to them via the framework, because they can't do anything with it: modulo [runtime state](https://github.com/hashicorp/terraform-plugin-framework/blob/main/docs/design/runtime-state.md#data), Create is a pure function of prior/planned resource state and config, to new state. 

Regarding the following:
```go
type Resource interface{
  GetSchema()
  NewResource() Resource
  Create()
  Read()
  Update()
  Delete()
}
```

It's already an anti-pattern trying to make an object's factory or constructor be a method on that object. I don't think we need another reason why `NewResource` needs to be separated out.

I like the [mix of structs and interfaces](https://github.com/hashicorp/terraform-plugin-framework/blob/main/docs/design/structs-interfaces.md#alternative-3-mix-of-structs-and-interfaces) approach, for reasons possibly different to the ones you go into here. The result might be the same, though:

```go
type ResourceFactory interface {
  NewResource(Provider) (Resource, []*tfprotov5.Diagnostic)
}

type Resource struct {
  Schema                *tfprotov5.Schema
  Implementation        func(p ProviderState) ResourceImplementation
}

type ResourceImplementation interface {
  Create(ctx context.Context, *CreateResourceRequest, *CreateResourceResponse)
  Read(ctx context.Context, *ReadResourceRequest, *ReadResourceResponse)
  Update(ctx context.Context, *UpdateResourceRequest, *UpdateResourceResponse)
  Delete(ctx context.Context, *DeleteResourceRequest, *DeleteResourceResponse)
}
```

The crux of my response is: What benefit would provider developers get from separating `ResourceType` and `ResourceValue`, except what they already get from separating a resource struct type from a resource interface from a resource factory (interface)?

Reply:
I think the `ResourceType`/`Resource` distinction is basically the mix of structs/interfaces approach you linked to just with different names and different contexts. If you think of `Resource` in the example you linked as `ResourceType`, and `ResourceImplementation` as `Resource`, you basically have the same exact types I documented above; you just don't need the `ResourceFactory` type for anything, arguably.

I think why I'm calling it out here is:

1. There's no guarantee we're going to choose to use that approach, so I want to address the trade-offs in this document, too. That way, if we decide to go with another approach, we can still talk about the trade-offs of separating or combining these concepts when we get to this decision and are considering it.
2. I think the struct and interface doc does a great job of examining how that choice impacts provider devs through its lens (the Go type system we expose), but I think it's out of scope for that document to talk about what I'm trying to address here, which is conceptual distinctions.

I'm gonna try to summarize the conceptual distinction I'm trying to draw, because I think I've been having difficulty articulating it, so I'm gonna take another stab at it. If you think you understand it and just disagree, please feel free to chime in with that and I'll stop trying to explain it. I just feel like I'm circling around what I'm trying to say and haven't quite found the center of the idea yet.

I think the core distinction I'm trying to make here is between an instance of a resource and the type of a resource. The protocol itself makes this distinction: the RPC requests define a [TypeName](https://pkg.go.dev/github.com/hashicorp/terraform-plugin-go/tfprotov6#ApplyResourceChangeRequest.TypeName) property identify the resource type they're operating on. The schema is returned using [a separate RPC](https://pkg.go.dev/github.com/hashicorp/terraform-plugin-go/tfprotov6#GetProviderSchema) which is asking about the resource type, not any specific instance of the resource. Core makes the distinction between a resource type and a resource instance. I think the question I'm asking here is: do we want to make that distinction, too? I don't know what my answer is, but I think it's a reasonable question to explore.

You raise that providers don't need to have the concept of resource instances exposed to them, because the function is pure and has all the data it needs (besides, presumably, the [provider-owned runtime data](https://github.com/hashicorp/terraform-plugin-framework/blob/main/docs/design/runtime-data.md)) passed via the arguments into it. I think this is true, but maybe not _obviously_ true; the type the `Create` method is defined on, arguably, is state that provider developers can store things in during Create or Read, and if we don't create a clear distinction or explicitly say it's not supported, they may see code that _sometimes_ works (subject to race conditions and graph ordering) and build a reliance on that.

I think one of the trade-offs I didn't list in this section, that I'll add when revamping it, is that if we make that distinction, we need to figure out how we're going to give provider developers access to that schema information in their CRUD methods (if we can't prove they don't need it).

Reply:
Also, I guess it's worth calling out, since we're making the decision about mixing structs/interfaces separately from this, if we decide to do that, and then need to consider this decision, I'd like this option explicitly called out so there's no chance we'll end up going with an option that doesn't appear in this document.

Reply:
Incomplete feedback to get started:

>I think the ResourceType/Resource distinction is basically the mix of structs/interfaces approach you linked to just with different names and different contexts.
Yes, this is what I meant by "I like the mix of structs and interfaces approach, for reasons possibly different to the ones you go into here. The result might be the same, though". Perhaps that doc would be a better place for the discussion of the conceptual distinction being drawn here, except for how it impacts factories in particular.

>the type the Create method is defined on, arguably, is state that provider developers can store things in during Create or Read, and if we don't create a clear distinction or explicitly say it's not supported, they may see code that sometimes works (subject to race conditions and graph ordering) and build a reliance on that.
The doc should be clear on the claims I think are being made here, that:
1. The provider-defined Resource struct type, which implements the framework.Resource interface, maps to the concept of a resource instance, and
2. This would sometimes work and sometimes not (this should be expanded on with examples of what "working" means).